include/oneapi/dpl/pstl/omp/util.h

```diff
@@ -153,6 +154,90 @@ __process_chunk(const __chunk_metrics& __metrics, _Iterator __base, _Index __chu
     __f(__first, __last);
 }
 
+// abstract class to allow inclusion in __thread_enumerable_storage as member without requiring explicit template
+// instantiation of param types
+template <typename _StorageType>
+class __construct_by_args_base
+{
+  public:
+    virtual ~__construct_by_args_base() {}
+    virtual std::unique_ptr<_StorageType> construct() = 0;
+};
+
+// Helper class to allow construction of _StorageType from a stored argument pack
+template <typename _StorageType, typename... _P>
+class __construct_by_args : public __construct_by_args_base<_StorageType>
+{
+  public:
+    std::unique_ptr<_StorageType>
+    construct() override
+    {
+        return std::move(
+            std::apply([](auto... __arg_pack) { return std::make_unique<_StorageType>(__arg_pack...); }, __pack));
+    }
+    __construct_by_args(_P&&... __args) : __pack(std::forward<_P>(__args)...) {}
+
+  private:
+    std::tuple<_P...> __pack;
+};
+
+template <typename _StorageType>
+struct __thread_enumerable_storage
+{
+    template <typename... Args>
+    __thread_enumerable_storage(Args&&... __args) : __num_elements(0)
+    {
+        __construct_helper =
+            std::make_unique<__construct_by_args<_StorageType, Args...>>(std::forward<Args>(__args)...);
+        _PSTL_PRAGMA(omp parallel)
+        _PSTL_PRAGMA(omp single) { __thread_specific_storage.resize(omp_get_num_threads()); }
+    }
+
+    std::uint32_t
+    size() const
+    {
+        // only count storage which has been instantiated
+        return __num_elements.load();
+    }
+
+    _StorageType&
+    get_with_id(std::uint32_t __i)
+    {
+        assert(__i < size());
+
+        std::uint32_t __count = 0;
+        std::uint32_t __j = 0;
+
+        for (; __j < __thread_specific_storage.size() && __count <= __i; ++__j)
+        {
+            // Only include storage from threads which have instantiated a storage object
+            if (__thread_specific_storage[__j])
+            {
+                __count++;
+            }
+        }
+        // Need to back up one once we have found a valid storage object
+        return *__thread_specific_storage[__j - 1];
+    }
+
+    _StorageType&
+    get_for_current_thread()
+    {
+        std::uint32_t __i = omp_get_thread_num();
+        if (!__thread_specific_storage[__i])
```

Comment:
What if we using `std::call_once` flag for every thread?
And declare it as `thread_local`...
In this case I believe we didn't need `__num_elements` anymore and may safe use `std::vector::size()` in the `size()` method. The justification - in the `__pattern_histogram` implementation we see two separate phases:
- in the first phase we create storages for each thread and in the second phase we simple accumulate data from them.

Reply:
We could do this, but it would be at a cost to performance, especially for small input sizes and for machines with many threads.  

For histogram at least, it is a large overhead cost to allocate, initialize and then finally accumulate the temporary storage.  So if we have an input size which would only warrant a few threads, we would rather not deal with the overhead for all threads.  This is why we prefer a TLS which only allocates as threads encounter the need for it.  

On the other hand, for small values of `n` our performance is already not very good for the "even bins" API. If we choose to abandon performance of the small case, this would certainly simplify the openMP implementation.

Reply:
ok